### ISSUE TITLE
OCI: Use manifest digest vs. tarball digest

### ISSUE THREAD

**Issue by jdolitsky** [helm/helm#8248]
For OCI artifacts, the top-level "envelope" of the artifact is the manifest. This is equally as unique as the chart tarball layer. Currently, on "helm chart save", "helm chart list", and others, the digest displayed is that of the chart tarball layer.

Here is some example output from `helm chart save`:
```
$ helm chart save mychart/ localhost:5000/myrepo/mychart:2.7.0
ref:     localhost:5000/myrepo/mychart:2.7.0
digest:  1b251d38cfe948dfc0a5745b7af5ca574ecb61e52aed10b19039db39af6e1617
size:    2.4 KiB
name:    mychart
version: 0.1.0
2.7.0: saved
```

The digest displayed here is `1b251d38cfe948dfc0a5745b7af5ca574ecb61e52aed10b19039db39af6e1617`. This is that of the chart tarball, represented by `layers[0].digest` in the manifest:

```
$ cat ~/Library/Caches/helm/registry/cache/blobs/sha256/31fb454efb3c69fafe53672598006790122269a1b3b458607dbe106aba7059ef | jq
{
  "schemaVersion": 2,
  "config": {
    "mediaType": "application/vnd.cncf.helm.config.v1+json",
    "digest": "sha256:8ec7c0f2f6860037c19b54c3cfbab48d9b4b21b485a93d87b64690fdb68c2111",
    "size": 117
  },
  "layers": [
    {
      "mediaType": "application/tar+gzip",
      "digest": "sha256:1b251d38cfe948dfc0a5745b7af5ca574ecb61e52aed10b19039db39af6e1617",
      "size": 2487
    }
  ]
}
```

Instead, the digest displayed should be `31fb454efb3c69fafe53672598006790122269a1b3b458607dbe106aba7059ef`, the digest of the manifest itself.

This also affects the output of `helm chart list` (note the DIGEST column, displaying first 7 characters of the digest):

```
$ helm chart list
REF                                                     NAME                    VERSION DIGEST  SIZE            CREATED
localhost:5000/myrepo/mychart:2.7.0                     mychart                 2.7.0   84059d7 454 B           27 seconds
localhost:5000/stable/acs-engine-autoscaler:2.2.2       acs-engine-autoscaler   2.2.2   d8d6762 4.3 KiB         2 hours
localhost:5000/stable/aerospike:0.2.1                   aerospike               0.2.1   4aff638 3.7 KiB         2 hours
localhost:5000/stable/airflow:0.13.0                    airflow                 0.13.0  c46cc43 28.1 KiB        2 hours
localhost:5000/stable/anchore-engine:0.10.0             anchore-engine          0.10.0  3f3dcd7 34.3 KiB 
```

From a registry's perspective, manifests are a "first-class citizen" which is used to further determine metadata and any included layers (such as the chart tarball layer).

The proposal here is to change this digest to use the manifest digest instead. This would normally be considered a breaking change, but the OCI feature set is still in experimental mode.

**Comment by mattfarina** [comment]
@jdolitsky Two questions for you...

1. When you request a chart by manifest, rather than tag, you need to use the manifest digest instead of the tar manifest, right?
2. The tar digest is the same as the digest stored in a provenance file, correct?

**Comment by jdolitsky** [comment]
@mattfarina -

> 1. When you request a chart by manifest, rather than tag, you need to use the manifest digest instead of the tar manifest, right?

Correct - however I'm not sure fetch by SHA is supported yet by Helm client (will have to verify)

> The tar digest is the same as the digest stored in a provenance file, correct?

Correct. 

On the topic of provenance files, these could potentially be stored as a second layer, and if present, the verification could occur on fetch. This would, however, modify the manifest digest. I'm wondering if provenance files should be ditched for OCI charts in favor of TUF. There is already a PR open for Notary v1 support: https://github.com/helm/helm/pull/7829

**Comment by mattfarina** [comment]
@jdolitsky I've also wondered about provenance files and TUF/Notary.

One of the features we have with Helm and provenance right now is that I can take a chart archive and the provenance file from the repo I got it from and move it elsewhere. For example, I can have my own mirror. The provenance will carry along with it because the signature has not changed and the digest of the archive file has not changed.

But, Notary has to do with signing a thing in it's current location. If something is moved to a new location, such as a mirror, the signing is lost. I wouldn't want to loose that feature in Helm. I think Notary v2 is working on that, though.

The reason I asked about the tar digest is because that's the digest we would want to check locally for the file downloaded. If we perform a hash of the tar archive downloaded we can compare it to the file digest to make sure we got what we expected. It helps to mitigate one type of attack against the software supply chain.
